### ISSUE TITLE
TORCH_DISABLE_SHARE_RDZV_TCP_STORE=0 is not compatible with torchelastic restarts

### ISSUE THREAD

**Issue by d4l3k** [pytorch/pytorch#135712]
### 🐛 Describe the bug

> Hi torch team, we recently upgraded our torch from 2.3 to 2.5.0dev0808. After the upgrade, we encountered issues with job automatic restart. The symptom is that during restart, the NCCL initialization would fail the error message was "Software caused connection issue", "Failed to connect to  <ip_addr>:<port>".
We did some debugging, and found the issue to be related to the sharing of TCP store  between torchrun's rendezvous handler and training process's init_process_group(). which was default behavior after commit bb13fad7aa7754042efe6e9465410cf5e543a77e, titled Share TCPStore by default when using c10d rdzv handler
It seems because of sharing TCP Store, the action of broadcastNCCLUniqueId does not work properly, sometimes it will get the uniqueID of previous restart, which resulted in wrong Port being used by NCCL's init.
We tried to disable the behavior of sharing TCP store by setting TORCH_DISABLE_SHARE_RDZV_TCP_STORE to 1, which seems to resolve the issue.

Root cause is the new shared rendezvous implementation behavior with restart counts. We use `TORCHELASTIC_RESTART_COUNT` to isolate each retry but this isn't globally consistent -- each worker keeps track of it locally so if some workers crash and restart it gets reset. This results in different workers using different `PrefixStores`.

https://github.com/pytorch/pytorch/blob/39a61795e3ee41eff4dfe76da14b2535cf47429b/torch/distributed/rendezvous.py#L186

There's a couple of options for fixing this:

1. disable shared tcp store (prior behavior -- has issues during shutdown)
2. add a new global counter that's actually consistent across all hosts (but not all rendezvous implementations support counts)
3. create a new tcpstore that is managed by elastic on every rendezvous (simplest)

For option 3: we actually already have the bulk of the code written for this -- we would just need to make it be the default case: https://github.com/pytorch/pytorch/blob/18a90309527e2685e0dafc916de2e17c086b679a/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py#L1192-L1209

### Versions

nightly

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @c-p-i-o

**Comment by wconstab** [comment]
is there any downside to (3)?

(1) is not ideal since as you mentioned, we moved to the shared server to avoid other issues during rank0 shutdown.  We guarantee the lifetime of TCPStore server is longer than the lifetime of any trainer, so trainers can assume it's safe to talk to TCPStore without having to worry about checking whether it may have shut down yet.

**Comment by fduwjj** [comment]
@wconstab if we only create this store once (I don't know if it is possible) then we can use it for a synchronization of different attemps across hosts?

**Comment by d4l3k** [comment]
@wconstab I don't think there's any big downsides to it -- there's a tiny-tiny risk that the host would run out of ephemeral ports but that would cause other bigger issues. 3 neatly avoids the free port issue as we bind it in process and since we start it from the rank0 host during rendezvous we won't have any issues with shutdown as rendezvous needs to happen first, and startup is just the normal startup path

@fduwjj what kind of synchronization are you looking for? We actually have other issues with this shared store such as the nccl dump signal ignores the prefix and writes to the root. Plus the shared store doesn't persist across retries when using gang scheduler retries

**Comment by fduwjj** [comment]
@d4l3k I mean for option 3, we only create the extra TCPStore once, and that store will be used for synchronization of the attempt. Or you are saying we completely ditch the shared TCPStore after "creating a new tcpstore that is managed by elastic on every rendezvous" ?

**Comment by wconstab** [comment]
i think iiuc, by destroying and recreating the store on each job attempt, we ensure that the new store starts with a fresh counter.  This is an effective (if a little hacky) way to ensure all trainers align on the same value of the new counter.  Option 2 would be to add a new counter key in software, to achieve the same thing - but at the cost of having to update various codepaths.  If that's all correct understanding, going with (3) feels fine to me.

**Comment by wzhang313** [comment]
> (1) is not ideal since as you mentioned, we moved to the shared server to avoid other issues during rank0 shutdown. 

@d4l3k 

Can you elaborate on the rank0 shutdown issue of (1)? We are currently using (1) as a workaround for this issue.
